scripts/patch-missing-contract-abis.ts

```diff
@@ -0,0 +1,121 @@
+  connectionConfig: getConnectionConfig(),
+});
+
+try {
+  console.log('Starting script to patch missing contract ABIs...');
+
+  // 2) Initialize script variables and RPC client
+  let lastBlockHeight = LAST_BLOCK_HEIGHT; // Initial value for the first query
+
+  let totalConsideredCount = 0;
+  let totalPatchedCount = 0;
+
+  const rpc = new StacksCoreRpcClient(); // Default to RPC host from ENV
+
+  // 3) Main processing loop: Fetch and patch contracts in batches
+  while (true) {
+    // 3.1) Find contracts whose ABI is still missing (paginated)
+    const missing = await sql<{ contract_id: string; block_height: number }[]>`
+      SELECT contract_id, block_height
+      FROM smart_contracts
+      WHERE (abi::text = '"null"')
+        AND canonical = TRUE
+        AND block_height > ${lastBlockHeight}
+      ORDER BY block_height ASC
+      LIMIT ${BATCH_SIZE}
+    `;
+
+    if (missing.length === 0) {
+      if (totalConsideredCount === 0) {
+        console.log('  - No contracts with missing ABI found.');
+      } else {
+        console.log(`  - Patched ${totalPatchedCount}/${totalConsideredCount} contracts.`);
+      }
+      break; // Exit the while loop
+    }
+
+    console.log(`- Found batch of ${missing.length} contracts with missing ABIs.`);
+
+    // 3.2) Process each contract in the current batch
+    for (const contract of missing) {
+      totalConsideredCount++;
+      const { contract_id, block_height } = contract;
+      const [address, name] = contract_id.split('.');
+      if (!address || !name) {
+        console.warn(`  - Skipping invalid contract id: ${contract_id}`);
+        continue;
+      }
+
+      try {
+        // 3.3) Fetch ABI from the connected Stacks node
+        const abi = await rpc.fetchJson<ClarityAbi>(`v2/contracts/interface/${address}/${name}`);
+
+        if (!abi || typeof abi !== 'object' || Object.keys(abi).length === 0) {
+          console.warn(`  - Skipping ${contract_id}. Fetched empty or invalid ABI.`);
+          continue;
+        }
+
+        // 3.4) Update row for this contract still missing an ABI
+        const rows = await sql`
+          UPDATE smart_contracts
+             SET abi = ${abi}
+           WHERE contract_id = ${contract_id}
+             AND (abi::text = '"null"')
+             AND canonical = TRUE
+        `;
+        if (rows.count === 0) {
+          console.warn(`  - Failed to patch ${contract_id}. No rows updated.`);
+          continue;
+        }
+        console.log(`  - Patched ABI for ${contract_id}`);
+        totalPatchedCount++;
+      } catch (err: any) {
+        let errorMessage = 'Unknown error';
+        if (err instanceof Error) {
+          errorMessage = err.message;
+        } else if (typeof err === 'string') {
+          errorMessage = err;
+        } else if (err && typeof err.message === 'string') {
+          errorMessage = err.message;
+        }
+        console.error(`  - Failed to patch ${contract_id}:`, errorMessage);
```

Comment:
nit: for simplicity you might want to either `throw new Error(message)` so it automatically goes to stderr with your custom message, or import `logger` and do `logger.error(err, "some message")` which will give you the full stack trace, etc.